[GeoMechanicsApplication] Nodal extrapolation for interface elements

avdg81 opened the pull request:
**📝 Description**
Extended the nodal extrapolator such that it also supports interface elements (line and surface ones).

**🆕 Changelog**
- The interface geometry can now return its mid-geometry through the member functions `pGetGeometryPart` and `GetGeometryPart` (the provided part index is not being used).
- Since `pGetGeometryPart` returns a shared pointer to the underlying mid-geometry, the data member that keeps the mid-geometry also had to be changed to a shared pointer.
- Added overrides for members `GetGeometryFamily` and `GetGeometryOrderType` for class template `InterfaceGeometry`, which forward the request to the underlying mid-geometry.
- Any interface element can now return a list of its integration points (regardless of the adopted integration scheme).
- Any interface element can now return its mid-geometry.
- When calculating the linear extrapolation matrix, we now only need to provide it with the element itself rather than the element's geometry and integration method. This change enables us to define a unified API that works for all elements (recall that the integration method could not be easily extended towards ones that are commonly used by interface elements).
- Added a utility function that returns the integration points of any element, including interface elements.
- Added a utility function that calculates the determinants of the Jacobian matrices of any element, including interface elements.
- Extended the nodal linear extrapolator such that it also supports any interface element.
- Made a utility function that generates nodes public.
- Added element setup functions that create linear and quadratic quadrilateral elements.
- Extended several existing unit tests to cover the newly added overrides for interface geometry (i.e. getting the geometry's family as well as its order).
- Added new unit tests to cover the proper construction of the mid-geometry of any interface geometry.
- The unit tests of the linear nodal extrapolator have been updated such that they work with elements rather than geometries.
- Added a helper function that checks whether two given lists of points are equal to within a provided tolerance on a pair-by-pair basis.
- Reuse the UBlas utility functions to create expected matrices and vectors.

avdg81 committed:
`ExtrapolationUtilities::CalculateNodalVectors` now receives an element rather than geometry, integration method and element ID

avdg81 committed:
Generating nodes has been made public

avdg81 committed:
Let the unit test use elements rather than geometries

Also added utility functions for creating 2D 4-noded elements and 2D 8-noded elements.

In addition, fixed the position of one node of the 20-noded hexahedron.

avdg81 committed:
When calculating the extrapolation matrix we now only require an element

avdg81 committed:
Added a utility function to get the integration points of an element

avdg81 committed:
Use a more generic way of evaluating the shape function values

This way also works for interface elements.

avdg81 committed:
Added a new utility function to evaluate the determinants of the Jacobian matrices at the integration points

avdg81 committed:
Removed a redundant function parameter

avdg81 committed:
Moved a declaration closer to its first usage

Also removed a redundant template parameter.

avdg81 committed:
Refactored creating the lower order geometry

Use the element family type as basis to decide which geometry to produce.

avdg81 committed:
Implemented `GetGeometryFamily` for interface geometries

It returns the geometry family of the underlying mid-geometry.

avdg81 committed:
Added a unit test for 4+4 noded interface geometry

avdg81 committed:
Added a unit test for 8+8 noded interface geometry

avdg81 committed:
Added creating lower order geometry of lines

avdg81 committed:
Implemented `GetGeometryOrderType` for interface geometry

It returns the order of the underlying mid-geometry.

avdg81 committed:
Removed unused variable

avdg81 committed:
Can now calculate the extrapolation matrix of an interface element

The corresponding unit test covers a line interface element.

avdg81 committed:
Made a member function private

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/experiment-with-extrapolation-for-interface-elements and fixed merge conflicts

avdg81 committed:
Added a comment that explains the use of an override

WPK4FEM committed:
Added triangular and quadrilateral interface extrapolation test.

WPK4FEM committed:
Used ublasUtilities::CreateMatrix to setup expected result matrices, i.s.o. streaming in results.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/experiment-with-extrapolation-for-interface-elements and resolved merge conflicts

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/nodal-extrapolation-for-interface-elements

avdg81 committed:
Several minor improvements

- Use `auto` when the type can be induced.
- Made more variables `const`.
- Moved some variables closer to their first usage.
- Added a comment to explain why the extrapolation matrix of an interface element needs to expanded after calculating it from the mid-geometry.
- Reformatted a few lines that are (deliberately) skipped by clang-format.
- Renamed a variable to comply with the Kratos Style guide.
- Eliminated a few intermediate variables.
- Removed a redundant blank line.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/nodal-extrapolation-for-interface-elements

avdg81 committed:
Added a unit test to check the mid-geometry of a 2+2 line interface geometry

Also added a helper function that checks whether to collections of points are equal within a given absolute tolerance.

avdg81 committed:
Added a unit test to check the mid-geometry of a 3+3 line interface geometry

avdg81 committed:
Added a unit test to check the mid-geometry of a 3+3 surface interface geometry

avdg81 committed:
Added a unit test to check the mid-geometry of a 6+6 surface interface geometry

avdg81 committed:
Added a unit test to check the mid-geometry of a 4+4 surface interface geometry

avdg81 committed:
Added a unit test to check the mid-geometry of a 8+8 surface interface geometry

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_utilities.h:
For completeness sake, I had tried to include `gtest/gtest.h` (for `ASSERT_EQ`), but for some unclear reason when building it on my machine, this file could not be found. And I don't understand why. Leaving it out seems to work, but I still feel a bit uneasy about this.

rfaasse replied:
Hmm, I think we need to try and fix this (since now it probably depends on the include order whether this will work or not)

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/linear_nodal_extrapolator.cpp:
```diff
@@ -13,54 +13,65 @@
+        auto tmp = Matrix{2 * result.size1(), result.size2(), 0.0};
+        GeoElementUtilities::AddMatrixAtPosition(result, tmp, 0, 0);
+        GeoElementUtilities::AddMatrixAtPosition(result, tmp, result.size1(), 0);
+        std::swap(result, tmp);
```
I think this can be simplified a bit (which also reduces the number of operations), since `resize` by default preserves the data (unless you specify preserve = false when calling the function)
```suggestion
        result.resize(2 * result.size1(), result.size2());
        GeoElementUtilities::AddMatrixAtPosition(result, tmp, result.size1(), 0);
```

avdg81 committed:
Forgot to change `std::make_unique` to `std::make_shared`

After changing the smart pointer type of `mMidGeometry`, I had forgotten to update the construction of the data member. Fixed that now.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_geometries/test_interface_geometry.cpp:
```diff
@@ -243,6 +256,77 @@ KRATOS_TEST_CASE_IN_SUITE(CreatingInterfaceWithThreeNodesThrows, KratosGeoMechan
+    const auto&    r_mid_geometry     = interface_geometry.GetGeometryPart(unused_part_index);
+
+    const auto expected_mid_points = std::vector<Point>{{-0.5, 0.1, 0.0}, {6.0, 0.1, 0.0}};
+    ExpectPointsAreNear(r_mid_geometry.Points(), expected_mid_points);
```
The information to quickly see whether these values are correct is 'hidden' within the `CreateTwoPlusTwoNoded2DLineInterfaceGeometry`. It's a bit of a balance between short tests and clarity, but I would argue in this case (and the following tests) that since the locations of the nodes are needed to understand the test, it's better to have that data within the test itself.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_geometries/test_interface_geometry.cpp:
```diff
@@ -243,6 +256,77 @@ KRATOS_TEST_CASE_IN_SUITE(CreatingInterfaceWithThreeNodesThrows, KratosGeoMechan
+    constexpr auto unused_part_index  = std::size_t{0};
+    const auto&    r_mid_geometry     = interface_geometry.GetGeometryPart(unused_part_index);
+
+    const auto expected_mid_points = std::vector<Point>{{0.0, 0.0, 0.25}, {1.0, 0.0, 0.25}, {1.0, 1.0, 0.25}};
```
I'm not sure if it matters, but in x/y there is no averaging (since the points are equal in xy). Same holds for the next tests

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_linear_nodal_extrapolator.cpp:
```diff
@@ -13,12 +13,10 @@
-#include "geometries/quadrilateral_2d_4.h"
-#include "geometries/quadrilateral_2d_8.h"
-#include "geometries/triangle_2d_3.h"
-#include "geometries/triangle_2d_6.h"
```
Nice, especially when we include the geo_mechanics_fast_suite_without_kernel.h in the future (and therefore don't include these geometries implicitly), this will save quite some compilation time

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_linear_nodal_extrapolator.cpp:
```diff
@@ -69,24 +61,41 @@ KRATOS_TEST_CASE_IN_SUITE(NodalExtrapolator_GivesCorrectExtrapolationMatrix_For2
+    const auto extrapolation_matrix = nodal_extrapolator.CalculateElementExtrapolationMatrix(*p_element);
+
+    // clang-format off
+    const auto expected_extrapolation_matrix = UblasUtilities::CreateMatrix({{1.0, 0.0},
```
It might be clarifying to add a comment that the matrix here is different than just 2x the one for the 2D2 line, due to the different integration scheme.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_linear_nodal_extrapolator.cpp:
```diff
@@ -95,49 +104,94 @@ KRATOS_TEST_CASE_IN_SUITE(NodalExtrapolator_GivesCorrectExtrapolationMatrix_For2
+    const auto extrapolation_matrix = nodal_extrapolator.CalculateElementExtrapolationMatrix(*p_element);
+
+    // clang-format off
+    const auto expected_extrapolation_matrix = UblasUtilities::CreateMatrix({{1.0, 0.0, 0.0},
```
Similar to the line, it might be interesting to explain that we basically get 2 instances of $I$ due to the Lobatto scheme

avdg81 committed:
Added a missing include

Also resolved an undesired dependency of the GeoMechanicsApplication core code on testing code. In addition, several Kratos expect macros have been replaced by GTest macros.

markelov208 commented on applications/GeoMechanicsApplication/test_setup_utilities/element_setup_utilities.cpp:
```diff
@@ -124,6 +109,16 @@ std::vector<Kratos::Point> ElementSetupUtilities::CreatePointsFor3D6NInterfaceEl
             {0.0, 0.0, 0.1}, {1.0, 0.0, 0.1}, {0.0, 1.0, 0.1}};
 }
 
+PointerVector<Node> ElementSetupUtilities::GenerateNodes(const std::vector<Point>& rPoints)
```
a very useful change

avdg81 committed:
Processed Richard's review suggestions

- The linear nodal extrapolator now only knows about geometry, and no longer about elements. The extrapolation utilities offer the entry point for calculating the extrapolation matrix for any kind of element, including interface elements.
- Simplified expanding the extrapolation matrix for interface elements.
- Defined the nodal positions in the test cases themselves to ease checking the expected values.
- Added an offset to the positions of the nodes at the second side, to check whether the mid-side nodes are properly positioned between both sides.
- Added a few clarifying comments.

avdg81 committed:
Processed Gennady's review suggestions

- Renamed a data member to comply with the Kratos Style Guide.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/nodal-extrapolation-for-interface-elements and fixed merge conflicts

Also renamed a function parameter for consistency reasons.

avdg81 committed:
Renamed another data member to comply with the Kratos Style Guide

avdg81 committed:
Added an element utility function to get the number of integration points

To avoid (too) many member functions for the interface elements, I have introduced a getter for the integration scheme. In turn, the integration scheme gives access to the number of integration points as well as the integration points themselves.

avdg81 committed:
Updated some `#include` lists

Also removed a few unused forward declarations.

avdg81 committed:
Cleaned up another `#include` list

avdg81 committed:
Cleaned up yet another `#include` list

markelov208 reviewed (approved):
Hi Anne, thank you very much for processing the comments. From my view-point, the PR is ready to be merged. Lets wait for Richard's decision.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/linear_nodal_extrapolator.h:
```diff
@@ -12,10 +12,10 @@
 #include "nodal_extrapolator.h"
-#include <cstddef>
+
+#include <memory>
```
Nice, thanks for the clean-up on includes
